server/src/main/java/org/elasticsearch/plugins/PluginsService.java

```diff
@@ -485,4 +544,163 @@ private static String signatureMessage(final Class<? extends Plugin> clazz) {
+        } else {
+            return parentLayers;
+        }
+    }
+
+    private static List<Configuration> parentConfigurationOrBoot(List<ModuleLayer> parentLayers) {
+        if (parentLayers == null || parentLayers.isEmpty()) {
+            return List.of(ModuleLayer.boot().configuration());
+        } else {
+            return parentLayers.stream().map(ModuleLayer::configuration).toList();
+        }
+    }
+
+    /** Ensures that the plugins main class (its entry point), if any, is accessible to the server. */
+    private static void ensureEntryPointAccessible(Controller controller, Module pluginModule, String className) {
+        if (className != null) {
+            controller.addOpens(pluginModule, toPackageName(className), serverModule);
+        }
+    }
+
+    /**
+     * Adds qualified exports declared in the server module descriptor to the target module.
+     * This is required since qualified exports targeting yet-to-be-created modules, i.e. plugins,
+     * are silently dropped when the boot layer is created.
+     */
+    private static void addQualifiedExports(Module target) {
+        serverModule.getDescriptor()
+            .exports()
+            .stream()
+            .filter(ModuleDescriptor.Exports::isQualified)
+            .filter(exports -> exports.targets().contains(target.getName()))
+            .forEach(exports -> serverModule.addExports(exports.source(), target));
+    }
+
+    /**
+     * Adds qualified opens declared in the server module descriptor to the target module.
+     * This is required since qualified opens targeting yet-to-be-created modules, i.e. plugins,
+     * are silently dropped when the boot layer is created.
+     */
+    private static void addQualifiedOpens(Module target) {
+        serverModule.getDescriptor()
+            .opens()
+            .stream()
+            .filter(ModuleDescriptor.Opens::isQualified)
+            .filter(opens -> opens.targets().contains(target.getName()))
+            .forEach(opens -> serverModule.addExports(opens.source(), target));
+    }
+
+    /**
+     * Tuple of module layer and loader.
+     * Modular Plugins are mapped to a single non-null layer and loader.
+     * Non-Modular plugins have a null layer.
+     */
+    record LayerAndLoader(ModuleLayer layer, ClassLoader loader) {
+
+        LayerAndLoader {
+            Objects.requireNonNull(loader);
+        }
+
+        static LayerAndLoader ofLoader(ClassLoader loader) {
+            return new LayerAndLoader(null, loader);
+        }
+    }
+
+    @SuppressForbidden(reason = "I need to convert URL's to Paths")
+    static final Path[] urlsToPaths(Set<URL> urls) {
+        return urls.stream().map(PluginsService::uncheckedToURI).map(PathUtils::get).toArray(Path[]::new);
+    }
+
+    static final URI uncheckedToURI(URL url) {
+        try {
+            return url.toURI();
+        } catch (URISyntaxException e) {
+            throw new UncheckedIOException(new IOException(e));
```

Comment:
Could this be an assertion error? These are files that it should never happen with right?

Reply:
Yes. Done.